instrumentation/network/build.gradle.kts

```diff
@@ -21,7 +21,7 @@ android {
 dependencies {
     api(platform(libs.opentelemetry.platform))
     api(libs.opentelemetry.api)
-    api(project(":common"))
+    api(project(":android-agent"))
```

Comment:
This contradicts my understanding of the original design:
<img width="731" alt="image" src="https://github.com/open-telemetry/opentelemetry-android/assets/75337021/2db2654b-6723-43e4-8440-84157f11016c">

It's odd to me that the specific instrumentation now depends on the topmost agent.  Is this something that goes away in part 3 or something?

Reply:
The general idea is to avoid the core parts of the agent having a dependency on the instrumentations. I agree it's confusing as it's shown here. My plan after all the instrumentations are created is to rename the existing `android-agent` module to something else, not sure yet, probably `core`, and then create a new `android-agent` module which will include the core along with all the instrumentations that we'd like the agent to have "enabled" out of the box.

Reply:
In the future, it would be nice for the Agent to not have an explicit dependency on the instrumentation modules. But that will require an interface layer using which instrumentations can be enabled/disabled/configured via the Agent.

Something something Gradle plugin for build time module inclusion/exclusion based on a config file :-)

Reply:
> In the future, it would be nice for the Agent to not have an explicit dependency on the instrumentation modules. But that will require an interface layer using which instrumentations can be enabled/disabled/configured via the Agent.

It's my understanding that this is literally what @LikeTheSalad is working on here -- by making the instrumentations implement the API, they can be service loaded via classpath at runtime, and the install API is lightweight and dependency free (at least from the agent to the instrumentation). It's happening.

Reply:
Yes, as @breedx-splk mentioned, avoiding the core part of the agent depending on the instrumentation modules is being done in these PRs. Though it's done in a way that, in the end, end users who add the `android-agent` dependency into their projects will still get all the current default instrumentations enabled because the final `android-agent` module will have a dependency on the instrumentations, alongside the "core" module. I did it this way to keep the same behavior somewhat unchanged after the refactoring (at least when it comes to adding the agent dependency into a project). What I understood from what @bidetofevil was mentioning is to use a different approach to include the default instrumentations in the future if I understood correctly? Maybe via a grade plugin that adds the instrumentation dependencies directly into the host project or the like. I haven't thought much about alternatives to add the default instrumentations tbh (these PRs are consuming most of my energy for now) though if we need further work that could keep on improving it I'm up to discuss it once I'm done with these PRs 😅 also because I think it's better not to add too many breaking changes at once.

Reply:
I think what I imagined the eventual state to be is what the diagram outlines - there being no direct or indirect dependency from the eventual Agent module on any module that contains instrumentation implementations. If that's what the eventual state will be, that's great - apologies for missing the context.

The second line is a completely different thought, and like @LikeTheSalad described, is more about the *App* using the Agent not having to explicitly include the modules in its gradle files, but rather have it automatically pulled in with a gradle plugin. This will allow us to have a central location to manage versioning instead of making the app update. But this is very much a forward-looking dream and not something that should be included in this PR or even in any short term plans 😅

Reply:
> I think what I imagined the eventual state to be is what the diagram outlines - there being no direct or indirect dependency from the eventual Agent module on any module that contains instrumentation implementations.

This is what will happen with the existing `android-agent` module, which at some point will be renamed to something like `core` or similar to denote the core/init functionality of the agent. Now, the _new_ `android-agent` module will have a dependency on `core` plus all the instrumentation modules that we want to provide by default. This is because of what we discussed some time ago about providing some instrumentations by default that we thought people would expect the agent to have out of the box. The only way I think we could avoid any dependencies on the agent, while still being able to provide default instrumentations, would be via a gradle plugin that adds those dependencies straight into the host project, though I agree that it sounds like something for a not-so-near future as I think we should discuss the pros/cons/alternatives in detail first.

For now, I'm trying to keep overall functionality somewhat intact, hence the module renaming to `core` to still provide a library named `android-agent` with the same default instrumentations as the current `android-agent` provides.

Reply:
Understood. Maybe we can discuss this further in the future - I believe we can make it so that eventually, the whole of the Agent co de do not need to know about or directly depending on any specific instrumentation's implementation code. But that won't be for right now and may not be worth it, so we can move on 😅